You are given a sorted integer array arr containing 1 and prime numbers, where all the integers of arr are unique. You are also given an integer k.

For every i and j where 0 <= i < j < arr.length, we consider the fraction arr[i] / arr[j].

Return the kth smallest fraction considered. Return your answer as an array of integers of size 2, where answer[0] == arr[i] and answer[1] == arr[j].

 

Example 1:

Input: arr = [1,2,3,5], k = 3
Output: [2,5]
Explanation: The fractions to be considered in sorted order are:
1/5, 1/3, 2/5, 1/2, 3/5, and 2/3.
The third fraction is 2/5.
Example 2:

Input: arr = [1,7], k = 1
Output: [1,7]
 

Constraints:

2 <= arr.length <= 1000
1 <= arr[i] <= 3 * 104
arr[0] == 1
arr[i] is a prime number for i > 0.
All the numbers of arr are unique and sorted in strictly increasing order.
1 <= k <= arr.length * (arr.length - 1) / 2
 

Follow up: Can you solve the problem with better than O(n2) complexity?

# approach:
- loop over the arr to find all the fractions.
- track the kth fraction using hte max-heap.

In [ ]:
import heapq

class Solution:
    def kthSmallestPrimeFraction(self, arr: list[int], k: int) -> list[int]:
        frac = []
        for i in range(len(arr)):
            for j in range(i+1, len(arr)):
                frac.append((-(arr[i]/arr[j]), arr[i], arr[j]))

        heapq.heapify(frac)
        while len(frac) > k:
           heapq.heappop(frac)

        print(frac)

        return [frac[0][1], frac[0][2]]
    
# tc :
# generating all fractions - O(n^2)
# heapify - O(n)
# kth fraction - O(k log k)

In [14]:
Solution().kthSmallestPrimeFraction(arr = [1,2,3,5], k = 3)

[(-0.4, 2, 5), (-0.3333333333333333, 1, 3), (-0.2, 1, 5)]


[2, 5]

In [ ]:
import heapq

class Solution:
    def kthSmallestPrimeFraction(self, arr: list[int], k: int) -> list[int]:
        frac = []
        for i in range(len(arr)):
            for j in range(i+1, len(arr)):
                frac.append((-(arr[i]/arr[j]), arr[i], arr[j]))

        max_heap = []
        for ele in frac:
            heapq.heappush(max_heap, ele)

            while len(max_heap) > k:
                heapq.heappop(max_heap)

        print(max_heap)

        return [max_heap[0][1], max_heap[0][2]]
    
# tc :
# generating all fractions - O(n^2)
# kth fraction - O(n log k)
# sc - O(n^2) + O(k)

In [18]:
Solution().kthSmallestPrimeFraction(arr = [1,2,3,5], k = 3)

[(-0.4, 2, 5), (-0.3333333333333333, 1, 3), (-0.2, 1, 5)]


[2, 5]

# Follow up: Can you solve the problem with better than O(n2) complexity?


# approach 2:
- we wanna compute only until the k elements, what we have to do is, intialize the heap with firs set of values.
- pop if its bi

In [21]:
# page 1: -heap


import heapq

class Solution:
    def kthSmallestPrimeFraction(self, arr: list[int], k: int) -> list[int]:
        n = len(arr)
        min_heap = []

        # Initialize heap with fractions of form (arr[0]/arr[j]) for j = 1 to n-1
        for j in range(1, n):
            heapq.heappush(min_heap, (arr[0] / arr[j], 0, j))

        # Pop k-1 smallest fractions from heap
        for _ in range(k - 1):
            value, i, j = heapq.heappop(min_heap)
            if i + 1 < j:
                # Add next fraction from the same denominator
                heapq.heappush(min_heap, (arr[i + 1] / arr[j], i + 1, j))

        # k-th smallest fraction is on top
        _, i, j = heapq.heappop(min_heap)
        return [arr[i], arr[j]]

# tc - O(k log k)
# sc - O(k)

Here’s the **complete dry run in a full table format** for:

> `arr = [1, 2, 3, 5]` and `k = 3`

We want the **3rd smallest prime fraction** formed by `arr[i]/arr[j]` where `i < j`.

---

### ✅ **Step-by-step Dry Run Table**

| Step | Operation | Fraction (arr\[i]/arr\[j]) | i | j | Value | Heap After Operation (Top 3)               |
| ---- | --------- | -------------------------- | - | - | ----- | ------------------------------------------ |
| 0    | Init Push | 1/2                        | 0 | 1 | 0.5   | \[(0.5, 0, 1)]                             |
| 0    | Init Push | 1/3                        | 0 | 2 | 0.333 | \[(0.333, 0, 2), (0.5, 0, 1)]              |
| 0    | Init Push | 1/5                        | 0 | 3 | 0.2   | \[(0.2, 0, 3), (0.5, 0, 1), (0.333, 0, 2)] |

---

| Step | Operation | Fraction | i | j | Value | Reasoning                                  |
| ---- | --------- | -------- | - | - | ----- | ------------------------------------------ |
| 1    | Pop       | 1/5      | 0 | 3 | 0.2   | 1st smallest fraction                      |
| 1    | Push      | 2/5      | 1 | 3 | 0.4   | i+1 = 1 < 3 → valid push                   |
|      | Heap      | —        | — | — | —     | \[(0.333, 0, 2), (0.5, 0, 1), (0.4, 1, 3)] |

---

| Step | Operation | Fraction | i | j | Value | Reasoning                                  |
| ---- | --------- | -------- | - | - | ----- | ------------------------------------------ |
| 2    | Pop       | 1/3      | 0 | 2 | 0.333 | 2nd smallest fraction                      |
| 2    | Push      | 2/3      | 1 | 2 | 0.666 | i+1 = 1 < 2 → valid push                   |
|      | Heap      | —        | — | — | —     | \[(0.4, 1, 3), (0.5, 0, 1), (0.666, 1, 2)] |

---

| Step | Operation | Fraction | i | j | Value | Reasoning                           |
| ---- | --------- | -------- | - | - | ----- | ----------------------------------- |
| 3    | Pop       | 2/5      | 1 | 3 | 0.4   | ✅ This is the 3rd smallest fraction |

---

### ✅ Final Result:

```python
[2, 5]
```



# most optimal: O(1) tc


In [ ]:
class Solution:
    def kthSmallestPrimeFraction(self, arr: list[int], k: int) -> list[int]:
        n = len(arr)
        low, high = 0.0, 1.0
        
        while high - low > 1e-9:
            mid = (low + high) / 2.0
            count = 0
            max_frac = 0.0
            numerator = 0
            denominator = 1
            i = -1  # pointer for numerator
            
            # Count the number of fractions < mid.
            #. we are checking line by line. count all the values < mid.
            # for each denominator
            for j in range(1, n):
                # increase i while the fraction is < mid
                while i + 1 < j and arr[i + 1] / arr[j] < mid:
                    i += 1
                
                # Count the number of value for this row.
                count += i + 1
                
                # track the max fraction ≤ mid
                if i >= 0 and arr[i] / arr[j] > max_frac:
                    max_frac = arr[i] / arr[j]
                    numerator = arr[i]
                    denominator = arr[j]
            
            if count < k:
                #if the fraction is < k. we need to make it big, go roght.
                low = mid
            else:
                # if >=, this could be our answer.
                # store it and move foraward.
                high = mid
                ans = [numerator, denominator]
        
        return ans

# O(n log(1/ε)) where:
# n is length of array
# ε is the precision (1e-9 here) :; maximum 60 iteration
# Each binary search step takes O(n) to count valid fractions using two pointers.
# sc - O(1)

0.6